src/Components/Components/src/RenderTree/RenderTreeBuilder.cs

```diff
@@ -590,13 +699,111 @@ public void Clear()
 
         private void Append(in RenderTreeFrame frame)
         {
+            var frameType = frame.FrameType;
             _entries.Append(frame);
 
-            var frameType = frame.FrameType;
             if (frameType != RenderTreeFrameType.Attribute)
             {
                 _lastNonAttributeFrameType = frame.FrameType;
             }
         }
+
+        // Internal for testing
+        internal void ProcessDuplicateAttributes(int first)
+        {
+            Debug.Assert(_hasSeenAddMultipleAttributes);
+
+            // When AddMultipleAttributes method has been called, we need to postprocess attributes while closing
+            // the element/component. However, we also don't know the end index we should look at because it
+            // will contain nested content.
+            var buffer = _entries.Buffer;
+            var last = _entries.Count - 1;
+
+            for (var i = first; i <= last; i++ )
+            {
+                if (buffer[i].FrameType != RenderTreeFrameType.Attribute)
+                {
+                    last = i - 1;
+                    break;
+                }
+            }
+
+            // Now that we've found the last attribute, we can iterate backwards and process duplicates.
+            var seenAttributeNames = (_seenAttributeNames ??= new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase));
```

Comment:
Could this be short-circuited if a component has no attributes i.e. `last` == `first`?

Reply:
Sure.

Reply:
This isn't super meaningful to optimize. We won't get into this code path unless someone called AddMultipleAttributes.